nova/virt/vmwareapi/vmops.py

```diff
@@ -3172,15 +3176,60 @@ def _sync_sync_server_group(context, sg_uuid):
                 LOG.debug('Sync for server-group %s done', sg_uuid)
                 return
 
+            # First we check for all instances, which have ongoing migrations
+            # on the given host, either as source or destination
+
+            # Decision matrix for migrations:
+            # Mig-Status Action/Host
+            #            Source  Dest
+            # Preparing  Remove  N/A
+            # Running    Remove  Add
+            # (Other states are consistent with default behaviour)
+            #
+            # So the Instance.host will always be the source of the migration,
+            # and we want to remove the rules.
+            # We only need to handle specially the case a running migration
+            # on the destination host, and add it
+
+            MigrationList = objects.migration.MigrationList
+            filters = {
+                "host": host,
+                "instance_uuid": sg.members,
+                "status": ["preparing", "running"],
+            }
+
+            expected_members = {}
+
+            migrations_by_instance_uuid = {}
+            for migration in MigrationList.get_by_filters(context, filters):
+                instance_uuid = migration.instance_uuid
+                if migration.source_compute == host:
+                    # The host is the source of a migration
+                    # That means the instance will be part of the instance list
+                    # So we have to remember that instance to be removed from
+                    # the DRS rule-set
+                    migrations_by_instance_uuid[instance_uuid] = \
+                        migration
+                else:
+                    # We now handle the destination side
+                    if migration.state == "preparing":
+                        # Not even started, we can ignore that one
+                        continue
+
+                    # Polling the cache, as vm_util.get_vm_ref is very slow
+                    # for the negative search.
+                    # As the live-migration will have
+                    moref = vm_util.vm_ref_cache_get(instance_uuid)
+                    if moref:
+                        expected_members[instance_uuid] = moref
```

Comment:
Should we maybe log if we expected to find a member but didn't find it? We do it in line 3251

Reply:
I am not sure what you mean by that. If we do not get a `moref`  then we do not expect it to be member of the instance.
And while the migration is still running, we do not know when to expect the instance to be there.

Reply:
As I read this, it is an error in case of a manual call during the migration, but can be ignored if the sync-loop or something else called us. Then a debug-log would still be nice, I guess, so we can see that we added the member during the manual call during the migration.

Reply:
Well, we only call it during a migration in the "post-migration"-phase, so it would fall though to the standard loop.
It is only to catch the case that the instance has already been migrated, while the db didn't "caught up" to the fact.
But now that we are talking about it, I am wondering if that is even necessary.

I am fearing a race condition between the sync-loop and the migration otherwise.

Reply:
Yes, I see the same racy behavior, but that does not mean, that we should not log if we didn't find something. It might be fine in most cases - as I said above -, but I would still log a debug-statement at least, so we know why the server-group was synced with less members than we might have expected at the time.

Is there a problem with adding a debug-log if we didn't find a moref, I'm overlooking?

Reply:
From my point of view, the default case is the other way around. We remove it likely already in the preparing state, and that is being logged.

The normal case is that the placement rules do not get added from a sync-call, while it is being migrated, but after the migration. 

However, should we "see" the VM already,  I want to avoid that the rules get removed, so I might add it prematurely before the migration itself adds the rules. I would rather log that.

But if you feel better about it, I can add log messages for both cases.

Reply:
Let's just merge. I don't feel too strongly. Thank you for explaining your view.